src/linker/Linker/DynamicDependency.cs

```diff
@@ -0,0 +1,140 @@
+
+#nullable enable
+
+namespace Mono.Linker
+{
+	/// Tracks dependencies created via DynamicDependencyAttribute in the linker.
+	/// This is almost identical to DynamicDependencyAttribute, but it holds a
+	/// TypeReference instead of a Type, and it has a reference to the original
+	/// CustomAttribute for dependency tracing. It is also a place for helper
+	/// methods related to the attribute.
+	internal class DynamicDependency : Attribute
+	{
+		public CustomAttribute? OriginalAttribute { get; set; }
+		public DynamicDependency (string memberSignature)
+		{
+			MemberSignature = memberSignature;
+		}
+
+		public DynamicDependency (string memberSignature, TypeReference type)
+		{
+			MemberSignature = memberSignature;
+			Type = type;
+		}
+
+		public DynamicDependency (string memberSignature, string typeName, string assemblyName)
+		{
+			MemberSignature = memberSignature;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, TypeReference type)
+		{
+			MemberTypes = memberTypes;
+			Type = type;
+		}
+
+		public DynamicDependency (DynamicallyAccessedMemberTypes memberTypes, string typeName, string assemblyName)
+		{
+			MemberTypes = memberTypes;
+			TypeName = typeName;
+			AssemblyName = assemblyName;
+		}
+
+		public string? MemberSignature { get; }
+
+		public DynamicallyAccessedMemberTypes MemberTypes { get; }
+
+		public TypeReference? Type { get; }
+
+		public string? TypeName { get; }
+
+		public string? AssemblyName { get; }
+
+		public string? Condition { get; set; }
+
+		public static DynamicDependency? ProcessAttribute (LinkContext context, ICustomAttributeProvider provider, CustomAttribute customAttribute)
+		{
+			if (!(provider is MethodDefinition || provider is FieldDefinition))
+				return null;
+
+			if (!DynamicDependency.ShouldProcess (context, customAttribute))
+				return null;
+
+			var dynamicDependency = GetDynamicDependency (context, customAttribute);
+			if (dynamicDependency != null)
+				return dynamicDependency;
+
+			context.LogMessage (MessageContainer.CreateWarningMessage ($"Invalid DynamicDependencyAttribute on '{provider}'", 2030));
+			return null;
+		}
+
+		static DynamicDependency? GetDynamicDependency (LinkContext context, CustomAttribute ca)
+		{
+			var args = ca.ConstructorArguments;
+			if (args.Count < 1 || args.Count > 3)
+				return null;
+
+			// First argument is string or DynamicallyAccessedMemberTypes
+			string? memberSignature = args[0].Value as string;
+			if (args.Count == 1)
+				return memberSignature == null ? null : new DynamicDependency (memberSignature);
+			DynamicallyAccessedMemberTypes? memberTypes = null;
+			if (memberSignature == null) {
+				var argType = args[0].Type;
+				if (!(argType.Namespace == "System.Diagnostics.CodeAnalysis" && argType.Name == "DynamicallyAccessedMemberTypes"))
+					return null;
+				try {
+					memberTypes = (DynamicallyAccessedMemberTypes) args[0].Value;
```

Comment:
This is wrong you are casting user type to linker compiled type

Reply:
This is enum - basically same story as with BindingFlags. Currently the type is defined in linker, once we switch over to .NET 5 it will come from the framework (on the linker side, just like BindingFlags do).

Reply:
I'm not sure this is going work like this. You probably need to cast it to int first as the type used inside linker does not have to much type in the linked assembly (basically there are 2 different mscorlibs/SPCs)

Reply:
You're right - I didn't realize that the runtime would do direct enum -> enum cast. We do need to go through integer.

Reply:
As long as the underlying type is the same, you can unbox into any enum you want.

```csharp
class Program
{
    enum SomeEnum { }
    enum OtherEnum : byte { }
    static void Main()
    {
        object o = (SomeEnum)123;
        _ = (OtherEnum)o;
    }
}
```

(I'm also pretty sure in this case the Value is a boxed integer, not an enum, but it works for that too.)

Reply:
The value is a boxed Int32 - so this is fine unless I misunderstood the concern.